InfluxDB component improvements

haimgel opened the pull request:
## Description:

This PR includes two improvements to InfluxDB component:

1. Allows reporting some state attributes as tags to InfluxDB. Some state attributes should really be tags in InfluxDB. E.g. it is helpful to be able to group by friendly_name, or add a custom attribute like "location" and group by that. Graphs in Grafana are much easier to read when friendly names are used, and not node ids.  This commit adds an optional setting to InfluxDB config: 'tags_attributes'. Any attribute on this list will be reported as tag and not as field to InfluxDB. 

2. Allow overriding InfluxDB measurement for each reported item separately. Bundling all items with the same "unit of measurement" together does not always makes sense. For example, both "relatively humidity" and "battery level" are reported as "%", but I'd rather see them as separate measurements in InfluxDB. This commit allows for 'influxdb_measurement' attribute. When set on node, it will take precedence over the global 'override_measurement' and component-specific 'unit_of_measurement'.

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3054

## Example entry for `configuration.yaml`:
```yaml

influxdb:
  tags_attributes: [friendly_name, location]

homeassistant:
  customize:
    sensor.master_bedroom_temperature:
      location: 'Master bedroom'
      influxdb_measurement: temperature

```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

haimgel committed:
Allow reporting some state attributes as tags to InfluxDB

Some state attributes should really be tags in InfluxDB. E.g.
it is helpful to be able to group by friendly_name, or add a custom
attribute like "location" and group by that. Graphs in Grafana are much
easier to read when friendly names are used, and not node ids.

This commit adds an optional setting to InfluxDB config:
'tags_attributes'. Any attribute on this list will be reported as tag
and not as field to InfluxDB.

haimgel committed:
Allow overriding InfluxDB measurement for each reported item separately

Bundling all items with the same "unit of measurement" together does not
always makes sense. For example, both "relatively humidity" and "battery
level" are reported as "%", but I'd rather see them as separate
measurements in InfluxDB. This commit allows for 'influxdb_measurement'
attribute. When set on node, it will take precedence over the global
'override_measurement' and component-specific 'unit_of_measurement'.

homeassistant commented:
Hi @hageltech, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#8633).

Once you do that we will be able to review and accept this pull request.

Thanks!

titilambert commented:
@hageltech @MartinHjelmare If this works as explained, I think this is really cool !

MartinHjelmare commented on homeassistant/components/influxdb.py:
```diff
@@ -25,7 +25,8 @@
 CONF_OVERRIDE_MEASUREMENT = 'override_measurement'
-CONF_BLACKLIST_DOMAINS = "blacklist_domains"
+CONF_INFLUXDB_MEASUREMENT = 'influxdb_measurement'
+CONF_TAGS_ATTRIBUTES = "tags_attributes"
```
Use single quotes here please.

MartinHjelmare commented on homeassistant/components/influxdb.py:
```diff
@@ -128,15 +132,17 @@ def influx_event_listener(event):
-                    measurement = default_measurement
-                else:
-                    measurement = state.entity_id
+        measurement = state.attributes.get(CONF_INFLUXDB_MEASUREMENT)
```
Add a comment before this line, explaining that this attributes can be set by a user in `customize`.

MartinHjelmare commented on homeassistant/components/influxdb.py:
```diff
@@ -153,7 +159,9 @@ def influx_event_listener(event):
-            if key != 'unit_of_measurement':
+            if key in tags_attributes:
+                json_body[0]['tags'][key] = value
+            elif not (key in ['unit_of_measurement', CONF_INFLUXDB_MEASUREMENT]):
```
```py
elif key not in ['unit_of_measurement', CONF_INFLUXDB_MEASUREMENT]:
```

MartinHjelmare commented on homeassistant/components/influxdb.py:
```diff
@@ -153,7 +159,9 @@ def influx_event_listener(event):
         for key, value in state.attributes.items():
-            if key != 'unit_of_measurement':
+            if key in tags_attributes:
+                json_body[0]['tags'][key] = value
```
I guess the user can handle any type mismatches in this case.

hageltech replied:
There shouldn't be any type mismatches since tags are always strings.

MartinHjelmare replied:
Yeah, but at least previously some types have still lead to problems. See https://github.com/home-assistant/home-assistant/issues/3657. At that time the influx library in use didn't handle/convert some types good. I'm not sure of the current state. Anyway, in this case the user can choose what attribute to record, so I don't think it will be a problem.

MartinHjelmare reviewed (changes requested):
Overall looks good. Some minor comments.

haimgel committed:
Minor updates to InfluxDB component improvements, as suggested by
@MartinHjelmare.

hageltech commented:
@MartinHjelmare thanks, I've added a commit with these small fixes.

pvizeli commented:
@balloob ? That is the first time who a component will try to access to a other abstraction layer. Normaly is that solve with a list inside this component config...

hageltech commented:
@pvizeli, I've considered doing it with the list inside the component config. However, I think keeping in inside customize is easier to maintain and less brittle. Things like this are possible:

```yaml
customize_glob:
  sensor.*humidity:
    influxdb_measurement: humidity
  sensor.*battery_level:
    influxdb_measurement: battery
```

I really think it's better than keeping essentially a list of all sensors inside influxdb config. 

Moreover, other things as defined in customize are of the same nature: emulated_hue, homebridge settings, etc.

MartinHjelmare reviewed (approved):
If @balloob is fine with the config in `customize` the rest looks good.

balloob commented:
> Moreover, other things as defined in customize are of the same nature: emulated_hue, homebridge settings, etc.

That will never be an argument to do / not do something. Best practices change over time.

I am not a fan of having things live in customize because it actually limits functionality. What if a user would want to have two InfluxDB components exporting to different databases? All of a sudden they will need the same config.

Another reason not to put it in customize is that it will become part of the representation of entities in the state machine. This means that it gets sent to the frontend and any other consumer of this data. None of them actually need this data.

So let's put it under the InfluxDB config. You can use the class that powers the customize with wildcards: [`EntityValues`](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity_values.py)

haimgel committed:
Moved per-component config from 'customize' into 'influxdb'
configuration section. The following three sub-sections were added:
'component_config', 'component_config_domain' and
'component_config_glob'. The sole supported per-component attribute
at this point is 'override_measurement'.

hageltech commented:
@balloob: fair enough.
Thanks for the EntityValues pointer. I've moved the config from 'customize' into 'influxdb' section.  I'll update the docs PR once this one gets approved.

balloob committed:
Lint

MartinHjelmare reviewed (changes requested):
We should add tests for the new config options. Forgot about that earlier, sorry.

titilambert commented:
@balloob having two different InfluxDB databases will limit thing you can do with Grafana.
But We can still have 2 influxdb components sending to the same database. 

BTW, I think the same is the Home Assistant attributes and InfluxDB format are not very compatible.
To improve that, HA attributes has to be more complex:
One attribute should be defined with a name, a value, a type (int, float, choice, ...) and a unit (and maybe some other stuff).
But I guess this kind of changes could really hard to implement.
Also, I'm an old monitoring (Nagios, Shinken, Zabbix, ...) guy, so maybe, my knowledge and experience are not suitable in home automation concepts.

titilambert commented:
@hageltech could you tell me when it's ready to test it ? If you want, I can try it on my current installation before the merge.

hageltech commented:
@titilambert It will be ready most probably tomorrow, or later tonight (EST). I'll add tests to new config options, and it seems current tests' mocked entity ids are incompatible with EntityValues, so that will have to be fixed as well.

haimgel committed:
Fixed mocked entity_ids in InfluxDB tests to be in domain.entity_id
format, to satisfy EntityValues requirements.

haimgel committed:
Added tests for new InfluxDB configuration parameters

hageltech commented:
@MartinHjelmare, I think the tests are good now.

MartinHjelmare commented on tests/components/test_influxdb.py:
```diff
@@ -529,3 +531,108 @@ def test_event_listener_default_measurement(self, mock_client):
             mock_client.return_value.write_points.reset_mock()
+
+    def test_event_listener_tags_attributes(self, mock_client):
+        """Test the event listener against a whitelist."""
```
Stale docstring.

MartinHjelmare commented on tests/components/test_influxdb.py:
```diff
@@ -529,3 +531,108 @@ def test_event_listener_default_measurement(self, mock_client):
+        mock_client.return_value.write_points.reset_mock()
+
+    def test_event_listener_component_override_measurement(self, mock_client):
+        """Test the event listener with a default measurement."""
```
Stale docstring.

MartinHjelmare reviewed (changes requested):
Two docstrings need updating. Otherwise :+1:

haimgel committed:
Fixes to some docstrings

MartinHjelmare reviewed (approved):
Nice!
